libbeat/processors/annotate/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,177 @@
+
+	config := kubeAnnotatorConfig{
+		InCluster:  true,
+		SyncPeriod: 1 * time.Second,
+		Namespace:  "kube-system",
+		LogsPath:   "/var/lib/docker/containers/",
+	}
+
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the annotate.kubernetes configuration: %s", err)
+	}
+
+	err = validate(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var c *rest.Config
+	if config.InCluster == true {
+		c, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to get in cluster configuration")
+		}
+	} else {
+		c, err = clientcmd.BuildConfigFromFlags(config.ApiServer, config.KubeConfig)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to get cluster configuration")
+		}
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(c)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create kubeclient due to error: %v", err)
+	}
+
+	if config.Host == "" {
+		podName := os.Getenv("HOSTNAME")
+		logp.Info("Using pod name %s and namespace %s", podName, config.Namespace)
+		if podName == "localhost" {
+			config.Host = "localhost"
+		} else {
+			pod, error := kubeClient.CoreV1().Pods(config.Namespace).Get(podName, meta_v1.GetOptions{})
+			if error != nil {
+				logp.Err("Querying for pod failed with error: ", error.Error())
+				logp.Info("Unable to find pod, setting host to localhost")
+				config.Host = "localhost"
+			} else {
+				config.Host = pod.Spec.NodeName
+			}
+
+		}
+	}
+
+	logp.Info("Using host ", config.Host)
+	logp.Info("Initializing watcher")
+
+	if kubeClient != nil {
+		watcher := NewPodWatcher(kubeClient, config.SyncPeriod, config.Host)
+		if watcher.Run() == true {
+			return kubernetesAnnotator{podWatcher: watcher, logsPath: config.LogsPath}, nil
+		} else {
+			return nil, fatalError
+		}
+
+	}
+
+	return nil, fatalError
+}
+
+func (k kubernetesAnnotator) Run(event common.MapStr) (common.MapStr, error) {
+
+	metadata := common.MapStr{}
+	if value, ok := event["source"]; ok {
+		source := value.(string)
+		logp.Debug("kubernetes", "Incoming source value: ", source)
+		cid := ""
+		if strings.Contains(source, k.logsPath) {
+			//Docker container is 64 chars in length
+			cid = source[len(k.logsPath) : len(k.logsPath)+64]
+		}
+		logp.Debug("kubernetes", "Using container id: ", cid)
+
+		if cid != "" {
+			metadata = k.podWatcher.GetMetaData(cid)
+		} else {
+			return event, nil
+		}
+	} else if value, ok := event["metricset-host"]; ok {
+		//This means that the event is from metricbeat/collectbeat
+		endpoint := value.(string)
+		logp.Debug("kubernetes", "using endpoint: ", endpoint)
+
+		if endpoint != "" {
+			metadata = k.podWatcher.GetMetaData(endpoint)
+		} else {
+			return event, nil
+		}
+
+	} else if value, ok := event["pod"]; ok {
+
+		logp.Debug("kubernetes", "using host pod name: ", value.(string))
+		metadata = k.podWatcher.GetMetaData(value.(string))
+
+		delete(event, "pod")
+	} else {
+		return event, nil
```

Comment:
i could have a filebeat configuration, that tails `/var/log/*` which is not managed by kubernetes and `/var/lib/docker/containers/*/*log` which is managed by kubernetes. the processor can be global and it would not pass through this check when a message from `/var/log` come in.